File: 240626/n개의 점 중 m개 고르기/choose-m-out-of-n-points.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <tuple>
#include <utility>
#include <climits>
using namespace std;

int n, m;
vector<pair<int, int>> spots;
vector<pair<int, int>> selected_spot;

int ans = INT_MAX;

int Dist(pair<int, int> a, pair<int, int> b) {
    int ax, ay;
    tie(ax, ay) = a;

    int bx, by;
    tie(bx, by) = b;

    return (ax-bx)*(ax-bx) + (ay-by)*(ay-by);
}

int Calc() {
    int max_dist = -1;
    for(int i=0; i<selected_spot.size(); i++) {
        for(int j=i+1; j<selected_spot.size(); j++) {
            max_dist = max(max_dist, Dist(selected_spot[i], selected_spot[j]));
        }
    }

    return max_dist;
}

void FindMinDist(int cnt, int cur_idx) {
    if(cnt == m) {
        ans = min(ans, Calc());
        return;
    }

    for(int i=cur_idx; i<spots.size(); i++) {
        selected_spot.push_back(spots[i]);
        FindMinDist(cnt+1, i+1);
        selected_spot.pop_back();
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    
    for(int i=0; i<n; i++) {
        int x, y;
        cin >> x>>y;
        spots.push_back(make_pair(x, y));
    }

    FindMinDist(0, 0);

    cout << ans;

    return 0;
}

File: 250926/N개의 점 중 M개 고르기/choose-m-out-of-n-points.py
import sys

INT_MAX = sys.maxsize

n, m = tuple(map(int, input().split()))

spots = [
    tuple(map(int, input().split()))
    for _ in range(n)
]

selc_spot = []
ans = INT_MAX

def calc():
    result = 0

    for i in range(m):
        for j in range(i+1, m):
            a, b = selc_spot[i]
            c, d = selc_spot[j]

            result = max(result, (a-c)**2 + (b-d)**2)
    
    return result

def find_min_dist(idx, cnt):
    global ans

    if cnt == m:
        ans = min(ans, calc())
        return
    
    if idx == n:
        return
    
    selc_spot.append(spots[idx])
    find_min_dist(idx+1, cnt+1)
    selc_spot.pop()

    find_min_dist(idx+1, cnt)


find_min_dist(0, 0)

print(ans)
